File: LeetCode-November-challenge/FindMedianfromDataStream.cpp
#include<bits/stdc++.h>
using namespace std;
class MedianFinder {
private:
    priority_queue<int> maxHeap;
    priority_queue<int, vector<int>, greater<int>> minHeap;
public:    
    void addNum(int num) {
        if(maxHeap.empty() && minHeap.empty()) maxHeap.push(num);
        else{
            if(maxHeap.top() < num) minHeap.push(num);
            else maxHeap.push(num);
        }
        int minSize = minHeap.size(), maxSize = maxHeap.size();
        if(minSize - maxSize == 2 || minSize - maxSize == -2){
            if(minSize > maxSize){
                int ele = minHeap.top();
                minHeap.pop();
                maxHeap.push(ele);
            }else{
                int ele = maxHeap.top();
                maxHeap.pop();
                minHeap.push(ele);
            }
        }
    }
    double findMedian() {
        int minSize = minHeap.size(), maxSize = maxHeap.size();
        if((minSize + maxSize) % 2 == 0)
            return (((double) minHeap.top()) + ((double) maxHeap.top())) / 2.0;
        if(minSize > maxSize) return minHeap.top();
        return maxHeap.top();
    }
};
int main(){
    MedianFinder *m = new MedianFinder();
    m->addNum(1);
    m->addNum(2);
    cout<<m->findMedian()<<endl;
    m->addNum(3);
    cout<<m->findMedian()<<endl;
}

File: Year-2022/LeetCode-November-challenge/FindMedianfromDataStream.java
import java.util.*;

class MedianFinder {
    private PriorityQueue<Integer> minHeap;
    private PriorityQueue<Integer> maxHeap;

    public MedianFinder() {
        minHeap = new PriorityQueue<>();
        maxHeap = new PriorityQueue<>(Collections.reverseOrder());
    }

    public void addNum(int num) {
        if (minHeap.isEmpty() && maxHeap.isEmpty())
            maxHeap.add(num);
        else {
            if (maxHeap.peek() < num)
                minHeap.add(num);
            else
                maxHeap.add(num);
        }
        int minSize = minHeap.size(), maxSize = maxHeap.size();
        if (Math.abs(minSize - maxSize) == 2) {
            if (minSize > maxSize) {
                int ele = minHeap.poll();
                maxHeap.add(ele);
            } else {
                int ele = maxHeap.poll();
                minHeap.add(ele);
            }
        }
    }

    public double findMedian() {
        int minSize = minHeap.size(), maxSize = maxHeap.size();
        if ((minSize + maxSize) % 2 == 0)
            return (((double) minHeap.peek()) + ((double) maxHeap.peek())) / 2.0;
        if (minSize > maxSize)
            return minHeap.peek();
        return maxHeap.peek();
    }
}

public class FindMedianfromDataStream {
    public static void main(String[] args) {
        MedianFinder m = new MedianFinder();
        m.addNum(1);
        m.addNum(2);
        System.out.println(m.findMedian());
        m.addNum(3);
        System.out.println(m.findMedian());
    }
}